internal/pkg/addon/package.go

```diff
@@ -0,0 +1,243 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package addon
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/aws/copilot-cli/internal/pkg/aws/s3"
+	"gopkg.in/yaml.v3"
+)
+
+type uploader interface {
+	Upload(bucket, key string, data io.Reader) (string, error)
+}
+
+// packagePropertyConfig defines how to package a particular property in a cloudformation resource.
+// There are two ways replacements occur. Given a resource configuration like:
+//  MyResource:
+//    Type: AWS::Resource::Type
+//    Properties:
+//      <Property>: file/path
+//
+// Without BucketNameProprty and ObjectKeyProperty, `file/path` is directly replaced with
+// the S3 location the contents were uploaded to, resulting in this:
+//  MyResource:
+//    Type: AWS::Resource::Type
+//    Properties:
+//      <Property>: s3://bucket/hash
+//
+// If BucketNameProperty and ObjectKeyProperty are set, the value of <Property> is changed to a map
+// with BucketNameProperty and ObjectKeyProperty as the keys.
+//  MyResource:
+//    Type: AWS::Resource::Type
+//    Properties:
+//      <Property>:
+//        <BucketNameProperty>: bucket
+//        <ObjectKeyProperty>: hash
+type packagePropertyConfig struct {
+	// Property is the key in a cloudformation resource's 'Properties' map to be packaged.
+	// Nested properties are represented by multiple keys in the slice, so the field
+	//  Properties:
+	//    Code:
+	//      S3: ./file-name
+	// is represented by []string{"Code", "S3"}.
+	Property []string
+
+	// BucketNameProperty represents the key in a submap of Property, created
+	// after uploading an asset to S3. If this and ObjectKeyProperty are empty,
+	// a submap will not be created and an S3 location URI will replace value of Property.
+	BucketNameProperty string
+
+	// ObjectKeyProperty represents the key in a submap of Property, created
+	// after uploading an asset to S3. If this and BucketNameProperty are empty,
+	// a submap will not be created and an S3 location URI will replace value of Property.
+	ObjectKeyProperty string
+
+	// ForceZip will force a zip file to be created even if the given file URI
+	// points to a file. Directories are always zipped.
+	ForceZip bool
```

Comment:
Two answers here:
1. We are not planning on adhering to that part of the spec - I agree with your concern.
2. `ForceZip: true` means that if you write an `AWS::Lambda::Function` with `Code: ./index.js`, we'll create a zip from that file instead of uploading it directly (i.e., to `s3://bucket/index.js`), because lambda requires s3 objects to be zip archives.